tagstudio/src/core/library/alchemy/library.py

```diff
@@ -0,0 +1,769 @@
+        with Session(self.engine) as session, session.begin():
+            # load all tags and join subtags
+            tags_query = select(Tag).options(selectinload(Tag.subtags))
+            tags = session.scalars(tags_query).unique()
+            tags_list = list(tags)
+
+            for tag in tags_list:
+                session.expunge(tag)
+                for subtag in tag.subtags:
+                    session.expunge(subtag)
+
+        return list(tags_list)
+
+    def save_library_to_disk(self):
+        logger.error("save_library_to_disk to be implemented")
+
+    def verify_ts_folders(self, library_dir: Path) -> None:
+        """Verify/create folders required by TagStudio."""
+        if library_dir is None:
+            raise ValueError("No path set.")
+
+        if not library_dir.exists():
+            raise ValueError("Invalid library directory.")
+
+        full_ts_path = library_dir / TS_FOLDER_NAME
+        if not full_ts_path.exists():
+            logger.info("creating library directory", dir=full_ts_path)
+            full_ts_path.mkdir(parents=True, exist_ok=True)
+
+    def verify_default_tags(self, tag_list: list) -> list:
+        """
+        Ensure that the default builtin tags  are present in the Library's
+        save file. Takes in and returns the tag dictionary from the JSON file.
+        """
+        missing: list = []
+
+        for m in missing:
+            tag_list.append(m)
+
+        return tag_list
+
+    def clear_internal_vars(self):
+        """Clear the internal variables of the Library object."""
+        self.library_dir = None
+        self.missing_files = []
+        self.dupe_files = []
+        self.ignored_extensions = []
+        self.missing_matches = {}
+
+    def refresh_dir(self) -> Iterator[int]:
+        """Scan a directory for files, and add those relative filenames to internal variables."""
+        if self.library_dir is None:
+            raise ValueError("No library path set.")
+
+        # Scans the directory for files, keeping track of:
+        #   - Total file count
+        start_time = time.time()
+        self.files_not_in_library: list[Path] = []
+        self.dir_file_count = 0
+
+        assert isinstance(self.library_dir, Path)
+        for path in self.library_dir.glob("**/*"):
+            str_path = str(path)
+            if (
+                path.is_dir()
+                or "$RECYCLE.BIN" in str_path
+                or TS_FOLDER_NAME in str_path
+                or "tagstudio_thumbs" in str_path
+            ):
+                continue
+
+            suffix = path.suffix.lower().lstrip(".")
+            if suffix in self.ignored_extensions:
+                continue
+
+            self.dir_file_count += 1
+            relative_path = path.relative_to(self.library_dir)
+            # TODO - load these in batch somehow
+            if not self.has_item(path=relative_path):
+                logger.info("item not in library yet", path=relative_path)
```

Comment:
This generates a lot of log spam when creating a library. Could this (and other messages) be moved to debug and just a summary logged to info?